sympy/physics/mechanics/kane.py

```diff
@@ -72,14 +76,18 @@ class KanesMethod(_Methods):
     configuration_constraints : iterable of Expr, optional
         Constraints on the system's configuration, i.e. holonomic constraints.
     u_dependent : iterable of dynamicsymbols, optional
-        Dependent generalized speeds.
+        Dependent generalized speeds. If complex constraints are used, observe
+        the comments further down.
     velocity_constraints : iterable of Expr, optional
         Constraints on the system's velocity, i.e. the combination of the
         nonholonomic constraints and the time-derivative of the holonomic
-        constraints.
+        constraints. They must be linear in q' and in u.
     acceleration_constraints : iterable of Expr, optional
         Constraints on the system's acceleration, by default these are the
         time-derivative of the velocity constraints.
+    complex_constraints : iterable of Expr, optional
+        They may be nonlinear in u and q', but must be linear in the time
+        derivatives of u and of q'.
```

Comment:
> You are already checking constraints for linearity, so is it necessary to have a new kwarg? If a user passed in nonlinear constraints to `velocity_constraints` then you can apply the new method.

Since I now use linear_eq_to_matrix a nonlinear error will be raised, but this could be handled, I suppose.
More important maybe: If a users by accident gives nonlinear constraints, but he intended to give linear ones, should he not be warned?
> 
> Also, what happens if both linear and nonlinear constraints are supplied? If we have a solution for nonlinear constraints, can that be applied to both linear and nonlinear constraints so that one solution works for both internally?

1. If it were a simple thing to merge them, why would Rothmayr explicitly treat them differently in his book?
2. If I treat linear motion constrains as if they were non-linear, I would differentiate them again w.r.t. time. Intuitively I#d loose accuracy (?)